### DISCUSSION TITLE
Unsupported Manifest/Layer Media type error

### DISCUSSION THREAD

**Discussion by corbinabeebe** [cdklabs/cdk-nextjs#39] [Q&A] [answered]
Hi All,

I received the following error:
```
Stack | 7:59:50 PM | CREATE_FAILED        | AWS::Lambda::Function                         | Nextjs/NextjsAssetsDeployment/Fn (NextjsNextjsAssetsDeploymentFn8972AAE9) 
Resource handler returned message: "The image manifest or layer media type for the source image XXXXX.dkr.ecr.us-west-2.amazonaws.com/cdk-XXXXX-container-assets-XXXXX-region:adksdlksaldsl is not supported. 
(Service: Lambda, Status Code: 400, Request ID: XXX)" (RequestToken: , HandlerErrorCode: InvalidRequest)

```
has anyone ran into this issue and what are the possible workarounds?

From what I can tell by the error, the Assets deployment Lambda function doesnt like the project image for some reason, do I need to build my image differently?
It would seem the image asset is not supported most likely, it looks like I can pass in my own dockerfile:
with add something like this:
```
          nextjsContainersProps:{
            dockerImageAsset: new aws_ecr_assets.DockerImageAsset(this, 'Image', {
              assetName: assetName, 
              directory: buildPath,
              buildSecrets: props.buildSecrets,
              }),
          },
```
Additionally, Is there a way I can define what ecr repo the image uploads too so I am not using the defualt repo for cdk?

Thanks!

**Comment by coltenkrauter** [comment]
I have not faced this...

Can you share your `Dockerfile`? The issue might be related to:

1. **Base Image**: Lambda works best with [Lambda-compatible base images](https://docs.aws.amazon.com/lambda/latest/dg/images-create.html), like `amazon/aws-lambda-nodejs:22`.
2. **Platform**: If you're using a custom image, ensure it's built for `linux/amd64`.

Example with [platform](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecr_assets.DockerImageAsset.html) specified:

```typescript
new aws_ecr_assets.DockerImageAsset(this, 'CustomImage', {
    directory: './path-to-dockerfile',
    platform: 'linux/amd64',
});

**Reply by corbinabeebe** [reply]
@coltenkrauter The error above I received when not passing in any custom dockerImageAsset props. I was wondering if maybe I need to store the image asset in a different ECR repo and how to do that.

I have also tried to pass in the image assets with taking the [global-containers.Dockerfile](https://github.com/cdklabs/cdk-nextjs/blob/3718f33c4cdd51702ad091b73d4d8176c1940b10/src/nextjs-build/global-containers.Dockerfile), but this errors out before I can even get to creating any resources on aws, I am most likely doing something completely wrong:
```
 - FromAsCasing: 'as' and 'FROM' keywords' casing do not match (line 4)
 - InvalidDefaultArgInFrom: Default value for ARG $BUILDER_IMAGE_TAG results in empty or invalid base image name (line 2)
Dockerfile:2
--------------------
   1 |     ARG BUILDER_IMAGE_TAG
   2 | >>> FROM $BUILDER_IMAGE_TAG as builder
   3 |     # Production image, copy all the files and run next
   4 |     FROM public.ecr.aws/docker/library/node:22-alpine as runner
--------------------
ERROR: failed to solve: base name ($BUILDER_IMAGE_TAG) should not be blank
```

**Reply by StuClift** [reply]
Lambda doesn't like the manifest docker produces. Consider setting https://docs.docker.com/build/building/variables/#buildx_no_default_attestations to true.

**Reply by coltenkrauter** [reply]
Ah I see. By chance, is your code in public repositories that you can share? I'd love to take a look. 

Any way you can share a bit more of the code might help in debugging. Meanwhile, @bestickley may have an idea of why you are experiencing this.

**Reply by corbinabeebe** [reply]
> Ah I see. By chance, is your code in public repositories that you can share? I'd love to take a look.
> 
> Any way you can share a bit more of the code might help in debugging. Meanwhile, @bestickley may have an idea of why you are experiencing this.

@coltenkrauter I can share more of the code but not the entire repo.  Anything in particular you would like me to share?

**Reply by corbinabeebe** [reply]
> Lambda doesn't like the manifest docker produces. Consider setting https://docs.docker.com/build/building/variables/#buildx_no_default_attestations to false.

How do I pass the arg in if I am not using a custom image asset?

**Reply by StuClift** [reply]
> How do I pass the arg in if I am not using a custom image asset?

`BUILDX_NO_DEFAULT_ATTESTATIONS=1 npm cdk deploy`

**Comment by bestickley** [comment, marked as answer]
@corbinabeebe , have you seen this? It offers some troubleshooting steps: https://repost.aws/questions/QUA20KN_o0SfOchiGO573GHw/cannot-deploy-docker-container-to-lambda

I've never seen that error.

> Additionally, Is there a way I can define what ecr repo the image uploads too so I am not using the defualt repo for cdk?

I haven't verified by this is what Amazon Q says:

```
Yes, you can customize the ECR repository name in your CDK code. For your NextjsRegionalContainers setup, you can specify this in the overrides:

const nextjs = new NextjsRegionalContainers(this, "Nextjs", {
  healthCheckPath: "/api/health",
  buildContext: fileURLToPath(new URL("..", import.meta.url)),
  overrides: {
    nextjsRegionalContainers: {
      nextjsBuildProps: {
        builderImageProps: {
          // Specify custom ECR repository name
          repositoryName: 'my-custom-repo-name',
          // ... your other settings ...
          exclude: [
            "**/node_modules",
            "**/.next",
            "global-containers",
            "global-functions",
            "regional-containers",
            "shared",
            "*.md",
          ],
        },
      },
    },
    // ... rest of your overrides ...
  },
  relativePathToWorkspace: "./app-playground",
});

Copy

Insert at cursor
typescript
Alternatively, if you want more control over the ECR repository, you can create it explicitly and then reference it:

import * as ecr from 'aws-cdk-lib/aws-ecr';

// In your stack:
const repository = new ecr.Repository(this, 'MyEcrRepo', {
  repositoryName: 'my-custom-repo-name',
  // Optional: Add lifecycle rules
  lifecycleRules: [
    {
      maxImageCount: 5,  // Keep only 5 images
      description: 'Only keep 5 images'
    }
  ],
  // Optional: Add removal policy
  removalPolicy: RemovalPolicy.DESTROY  // Be careful with this in production
});

const nextjs = new NextjsRegionalContainers(this, "Nextjs", {
  // ... other config ...
  overrides: {
    nextjsRegionalContainers: {
      nextjsBuildProps: {
        builderImageProps: {
          repository: repository,
          // ... other builder props ...
        },
      },
    },
  },
});

Copy

Insert at cursor
typescript
This gives you more control over the ECR repository configuration, including:

Custom repository name

Lifecycle rules

Image tag mutability

Image scanning configuration

Removal policies

Remember that if you're using a custom repository name, ensure you have the necessary IAM permissions set up correctly for pushing and pulling images from this repository.
```

**Reply by bestickley** [reply]
Marking this as answer for house keeping purposes. Please let me know if this doesn't address your issue.
